### PR TITLE
Implement a kclient package for kdo

### PR DESCRIPTION
**Pull request by johnmcollier**
Signed-off-by: John Collier <John.J.Collier@ibm.com>

## What is the purpose of this change? What does it change?
* Implements a `kclient` package for kdo, containing a number of useful kubernetes functions to be used throughout kdo. This will be required for us to start implementing a number of commands in kdo
    * We should be adding additional functions to this package as needed
* Brings over more unit tests from odo: 
    * kclient_test
    * validation_test
    * validators_test
    * preference_test
    * configs_test
* Removes dependencies on `odo` and OpenShift specific packages
    * github.com/openshift/api
    * github.com/openshift/client-go
    * github.com/openshift/odo
    * github.com/redhat-developer/odo

## How to test changes?
* Verify `kdo` builds with `make` or `make install`
* Run the current and new tests with `make test`

### CODE REVIEW

**Comment by johnmcollier** [comment]
@rajivnathan Please review when you get a chance

File: pkg/kclient/kclient.go [hunk 1]
```diff
@@ -0,0 +1,1313 @@
...
+
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	v1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/remotecommand"
+)
+
+var (
+	DEPLOYMENT_CONFIG_NOT_FOUND_ERROR_STR string = "deploymentconfigs.apps.openshift.io \"%s\" not found"
```
Comment (rajivnathan) [inline pkg/kclient/kclient.go]: Should this be changed to the generic kube deployment resource?

File: pkg/kclient/kclient.go [hunk 2]
```diff
@@ -0,0 +1,1313 @@
...
+
+// Define a function that is meant to create patch based on the contents of the DC
+type dcPatchProvider func(dc *appsv1.Deployment) (string, error)
+
+// this function will look up the appropriate Deployment, and execute the specified patch
+// the whole point of using patch is to avoid race conditions where we try to update
+// dc while it's being simultaneously updated from another source (for example Kubernetes itself)
+// this will result in the triggering of a redeployment
+func (c *Client) patchDepOfComponent(componentName, applicationName string, dcPatchProvider dcPatchProvider) error {
+	depName, err := util.NamespaceOpenShiftObject(componentName, applicationName)
```
Comment (rajivnathan) [inline pkg/kclient/kclient.go]: Should `NamespaceOpenShiftObject` be renamed to something more generic as well?

File: pkg/kclient/kclient.go [hunk 3]
```diff
@@ -0,0 +1,1313 @@
...
+
+	return dc.Spec.Template.Spec.Containers[0].Env, nil
+}
+
+// PropagateDeletes deletes the watch detected deleted files from remote component pod from each of the paths in passed s2iPaths
+// Parameters:
+//	targetPodName: Name of component pod
+//	delSrcRelPaths: Paths to be deleted on the remote pod relative to component source base path ex: Compoent src: /abc/src, file deleted: abc/src/foo.lang => relative path: foo.lang
+//	s2iPaths: Slice of all s2i paths -- deployment dir, destination dir, working dir, etc..
+func (c *Client) PropagateDeletes(targetPodName string, delSrcRelPaths []string, s2iPaths []string) error {
```
Comment (rajivnathan) [inline pkg/kclient/kclient.go]: This function seems to have some s2i specifics, should it be removed?

**Pull request merged**